### PR TITLE
[BEAM-11907] SQS source with full functionality

### PR DESCRIPTION
**Pull request by dennisylyung**
The current SqsIO source connector has limited functionality. 
This PR implements the following functionalities to SqsIO:
- Batch message deletion (mitigate throttled deleteMessage requests)
- Extension of visibility timeout (The existing implementation causes message duplication)
- Update watermarks when no new messages are pulled
New functionalities are written with reference to the pubsub connector. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by boyuanzz** [comment]
cc: @boyuanzz

**Comment by aaltay** [comment]
@dennisylyung - Thank you for your contribution. Is this PR ready for review?

**Comment by dennisylyung** [comment]
> @dennisylyung - Thank you for your contribution. Is this PR ready for review?

Not yet sorry.
I am implementing for the aws2 module too.

**Comment by dennisylyung** [comment]
Run Java PreCommit

**Comment by dennisylyung** [comment]
Run Java PreCommit

**Comment by dennisylyung** [comment]
R: @aaltay
It is ready for review now, thanks!
The PR turned out to be more rewriting the connectors referencing the pubsub one than doing small enhancements, since when I was developing a job with the sqs reader I found that it lacks many functionalities to make it generally usable. 
Please let me know how to properly document such changes.

**Comment by aaltay** [comment]
@dennisylyung - Thank you!

Not sure who might be a good reviewer for this PR. Do you mind pinging the dev@ thread you started (or a new thread) ask for reviews?

**Comment by dennisylyung** [comment]
@iemejia Do you think you can review, or ping someone who is suitable? Thanks!

**Comment by iemejia** [comment]
@dennisylyung I am a bit busy this week but maybe @aromanenko-dev can take a look. In the main time can you please rebase and squash your commits. Thanks !

**Comment by dennisylyung** [comment]
Run Java PreCommit

**Comment by dennisylyung** [comment]
Run Java PreCommit

**Comment by aromanenko-dev** [comment]
@iemejia Sure, I'll take a look as soon as I can (I hope in a couple of next days).

**Comment by aromanenko-dev** [comment]
In the same time, I see that these 2 gradle tasks failed:
`:sdks:java:io:amazon-web-services:compileJava`
`:sdks:java:io:amazon-web-services2:compileJava`

More details are [here](https://ci-beam.apache.org/job/beam_PreCommit_Java_Commit/17309/consoleFull)

**Comment by aromanenko-dev** [comment]
Run Java PreCommit

**Comment by dennisylyung** [comment]
Run Java PreCommit

**Comment by dennisylyung** [comment]
Run Java_Examples_Dataflow_Java11 PreCommit

**Comment by dennisylyung** [comment]
@aromanenko-dev It passes the java precommit tests now

**Comment by aromanenko-dev** [comment]
@dennisylyung Thanks! I'll take a look in the next week.

**Comment by aaltay** [comment]
> @dennisylyung Thanks! I'll take a look in the next week.

@aromanenko-dev, thank you and a gentle ping to get this in your review queue.

File: sdks/java/io/amazon-web-services/src/test/java/org/apache/beam/sdk/io/aws/sqs/SqsIOTest.java [hunk 1]
```diff
@@ -46,25 +52,151 @@
...
+      client.sendMessage(queueUrl, message);
+    }
+    source =
+        new SqsUnboundedSource(
+            SqsIO.read().withQueueUrl(queueUrl).withMaxNumRecords(1),
+            new SqsConfiguration(pipeline.getOptions().as(AwsOptions.class)));
+  }
+
+  @Test
+  public void checkpointCoderIsSane() {
```
[the remark this answers is not in the document]
Comment (dennisylyung) [reply sdks/java/io/amazon-web-services/src/test/java/org/apache/beam/sdk/io/aws/sqs/SqsIOTest.java]: changed in the new commit

File: sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/sqs/SqsUnboundedReader.java [hunk 2]
```diff
@@ -17,51 +17,385 @@
...
 })
-class SqsUnboundedReader extends UnboundedSource.UnboundedReader<Message> implements Serializable {
+class SqsUnboundedReader extends UnboundedSource.UnboundedReader<Message> {
+  private static final Logger LOG = LoggerFactory.getLogger(SqsUnboundedReader.class);
 
+  /** Maximum number of messages to pull from SQS per request. */
   public static final int MAX_NUMBER_OF_MESSAGES = 10;
+
+  /** Maximum times to retry batch SQS operations upon partial success. */
+  private static final int BATCH_OPERATION_MAX_RETIRES = 5;
```
[the remark this answers is not in the document]
Comment (dennisylyung) [reply sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/sqs/SqsUnboundedReader.java:82]: To be honest most are copied from the pubsub connector, except the update and delete batch size of 10, which is the limit of the AWS API.

**Review by aromanenko-dev — CHANGES_REQUESTED** [review]
It looks fine for me in general (since new functionality is mostly based on PubsubReader ones). I have just a minor request about tests.

@aaltay Please, ping someone who knows this PubsubReader functionality and who can review this in more details.

**Comment by aaltay** [comment]
@dpcollins-google / @boyuanzz - Would one of you also review this change? (@aromanenko-dev mentioned that most of the change is based from the pubsub io.)

**Comment by dpcollins-google** [comment]
@boyuanzz is probably a better reviewer, but I can if they're too busy. Reach out to me internally if you need my review

**Comment by aaltay** [comment]
@boyuanzz - Would you be able to review this when you get a chance?

**Comment by dennisylyung** [comment]
Run Java PreCommit

**Comment by dennisylyung** [comment]
Run Java PreCommit

File: sdks/java/io/amazon-web-services/src/test/java/org/apache/beam/sdk/io/aws/sqs/SqsIOTest.java [hunk 3]
```diff
@@ -46,25 +52,151 @@
...
+    CoderProperties.coderSerializable(source.getCheckpointMarkCoder());
+    // Since we only serialize/deserialize the 'notYetReadIds', and we don't want to make
+    // equals on checkpoints ignore those fields, we'll test serialization and deserialization
+    // of checkpoints in multipleReaders below.
+  }
 
-    final PCollection<Message> output =
-        pipeline.apply(SqsIO.read().withQueueUrl(queueUrl).withMaxNumRecords(100));
+  @Test
+  public void testReadOneMessage() throws IOException {
```
[the remark this answers is not in the document]
Comment (dennisylyung) [reply sdks/java/io/amazon-web-services/src/test/java/org/apache/beam/sdk/io/aws/sqs/SqsIOTest.java]: changed in the new commit

File: sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/sqs/SqsIOTest.java [hunk 4]
```diff
@@ -17,59 +17,200 @@
...
+  public void testCheckpointCoderIsSane() {
+    setupOneMessage();
+    CoderProperties.coderSerializable(source.getCheckpointMarkCoder());
+    // Since we only serialize/deserialize the 'notYetReadIds', and we don't want to make
+    // equals on checkpoints ignore those fields, we'll test serialization and deserialization
+    // of checkpoints in multipleReaders below.
+  }
 
+  @Test
+  public void testReadOneMessage() throws IOException {
```
[the remark this answers is not in the document]
Comment (dennisylyung) [reply sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/sqs/SqsIOTest.java]: changed in the new commit

**Comment by dennisylyung** [comment]
@aromanenko-dev I have separated the tests as suggested in your last review. 
However, the java tests seem to fail on other modules. Any idea why?

**Comment by aromanenko-dev** [comment]
@dennisylyung I don't think it's related to your change since only `org.apache.beam.sdk.io.elasticsearch.ElasticsearchIOTest.testWriteScriptedUpsert` fails. 
Though, it can be related to this [discussion](https://lists.apache.org/thread.html/r3979180772ff2166a95ff65735ff353a71dd42b2be16b643529b89c3%40%3Cdev.beam.apache.org%3E).

**Comment by aromanenko-dev** [comment]
Run Java PreCommit

**Review by aromanenko-dev — APPROVED** [review]
If there are no principal objections then I think it can be merged.

**Comment by aaltay** [comment]
> If there are no principal objections then I think it can be merged.

I agree. There is one approval, and all tests pass. Feel free to merge.

**Pull request merged**